Add load balancer

hardyoyo opened the pull request:
Add an elastic load balancer (and any required security groups) which would act as a reverse proxy, so a single domain  (associated with the load balancer) can support a DSpace instance. Includes:

- listens on port 443, and directs url patterns /server/* and /server to port 8080 on the api instance, and default     redirects (everything else) to port 443 on the client instance.
- listens on port 80, and redirects to HTTPS://#{host}:443/#{path}?#{query} with a status code of HTTP_301

**NOTE: this is a draft and should not be merged until the draft designation is lifted.**

Still to do:

[x] associate the passed Elastic IP address with the load balancer
[x] use the as-yet unused paramters (currently only have them in the metadata so cf-lint will keep quiet while I'm working)
[x] review the security groups template to ensure they are complete and correct, particularly the names and tags
[x] consider adding DependsOn constraints to any resources that depend on other resources (discussed with @bibliotechy, we've agreed we'll start out by trusting that Cloudformation will "get it right" and will only add explicit dependson constraints if it turns out they're necessary)

hardyoyo committed:
draft plan with TODOs for ALB

hardyoyo committed:
Add parameters to dspace-alb, Add draft security groups template

hardyoyo committed:
Rough in the load balancer, cfn-lint throws an error one line still

- line 107: E3002: Invalid Property
  Resources/ListenerHttps/Properties/Rules
- this sounds fishy to me, I may want to suppress E3002 for cfn-lint

hardyoyo committed:
ignoring unused parameters by, erm, using them, in metadata, pipe down cf-lint

hardyoyo committed:
Use AWS::ElasticLoadBalancingV2::ListenerRule for redirects

hardyoyo committed:
Add basic instructions to README

- Very much a work in progress, and not tested at all.
- Still not sure how to associate the ALB with an elastic IP

hardyoyo committed:
remove earlier draft of the load balancer template

hardyoyo committed:
udpate TODOs in dspace-alb.yaml

bibliotechy commented on dspace-alb.yaml:
```diff
@@ -1 +1,158 @@
+
+        ### remaining TODOs ####################################################
+        ### TODO: Add health checks
+        ### TODO: Add tags 
```
👏 tags 👏

bibliotechy commented on dspace-alb.yaml:
```diff
@@ -1 +1,158 @@
+    LoadBalancer:
+      Type: AWS::ElasticLoadBalancingV2::LoadBalancer
+      Properties:
+        Name: !Ref LoadBalancerName
```
Is there a specific reason to have the name passed in at runtime? I feel like we'd like to keep a consistent naming pattern across dev / stage / prod which is harder to enforce if we pass it in here, right?

hardyoyo replied:
it's a placeholder, I intend to build a name out of the other parameters

bibliotechy commented on dspace-alb.yaml:
```diff
@@ -1 +1,158 @@
+      Properties:
+        Name: !Join [ "-", [ !Ref LoadBalancerName, "client-target-group" ] ]
+        Port: 443
+        Protocol: HTTPS
```
If the LB is communicating with the backend service in a private network, do we need this over https? How are we managing the SSL cert (and domain) for that?

hardyoyo replied:
Good point. The current load balancer for the pilot instance in the sandbox is set up in this way.

hardyoyo replied:
we should probably talk about the SSL concerns, I think I understand what you're saying, but I want to be sure before I change anything.

bibliotechy commented on dspace-sg.yaml:
```diff
@@ -0,0 +1,80 @@
+      Type: AWS::EC2::SecurityGroup
+      Properties:
+        VpcId: !Ref VpcId
+        GroupName: LoadBalancerSecurityGroup
```
Please be more specific in the name. We should probably define a specific naming strategy. Something like `pad-dspace-lb-sg-{env}`.

hardyoyo replied:
Good suggestion

bibliotechy commented on dspace-sg.yaml:
```diff
@@ -0,0 +1,80 @@
+      Type: AWS::EC2::SecurityGroup
+      Properties:
+        VpcId: !Ref VpcId
+        GroupName: ApiInstanceSecurityGroup
```
MOre name specificity

bibliotechy commented on dspace-sg.yaml:
```diff
@@ -0,0 +1,80 @@
+        GroupName: ClientInstanceSecurityGroup
+        GroupDescription: Security group for the client instance
+        SecurityGroupIngress:
+          - IpProtocol: tcp
```
Again the question about managing SSL for backend when SSL termination is now happening at the LB

bibliotechy commented on dspace-sg.yaml:
```diff
@@ -0,0 +1,80 @@
+      Type: AWS::EC2::SecurityGroup
+      Properties:
+        VpcId: !Ref VpcId
+        GroupName: ClientInstanceSecurityGroup
```
Name specificity

bibliotechy reviewed (changes requested):
@hardyoyo Looks good, but I have some feedback naming and questions about SSL on backend when SSL termination is now happening at the LB

hardyoyo committed:
Add CDL paraeters, build the LoadBalancer Name from CDL parameters, drop the LoadBalancerName placeholder

hardyoyo committed:
WIP: use unused parameters, add EIP documentation to README

hardyoyo committed:
Improve naming and tagging in all templates, and other improvements

- Better names and tags, as suggested by @bibliotechy
- Ensure all EC2s are in the same subnets as the load balancer
- Ensure all security groups are used throughout

hardyoyo committed:
Fix double-SSL issues

- Move DSpace client target group (on the load balancer) and security group to HTTP/port 80
- Ensure health check for the client also uses port 80

hardyoyo committed:
update healthcheck paths to point towards working paths on each EC2 instance

hardyoyo committed:
remove TODO from alb template

hardyoyo committed:
Use a better HealthCheckPath for the API instance, as per @tdonohue's suggestion
